### PR TITLE
Support PUBLIC_ environment variables

### PR DESCRIPTION
**Pull request by jonathantneal**
## Changes

This PR adds support for exposing `PUBLIC_` environment variables to Astro. Environment variables starting with `PUBLIC_` will be exposed to client source code via `import.meta.env`.

## PUBLIC_ vs X_

The `PUBLIC_` prefix is more intuitive than the current `VITE_` prefix or a potential `ASTRO_` prefix. The `PUBLIC_` prefix is also more terse than `SNOWPACK_PUBLIC_` or `ASTRO_PUBLIC_`. The `PUBLIC_` prefix is an [established convention](https://documentation.help/HTMLREF/defaultiface.htm).

## Testing

A test for `.env` environment variables is added for both public and private values.

## Docs

Documentation in #1751.

### CODE REVIEW

**Review by natemoo-re — COMMENTED** [review]
I think this is great! I'd much prefer this interoperable convention over proliferating tool-specific prefixes.

One question: how _would_ you reference `SECRET_PLACE` if it were a DB key or something needed for the server?

File: packages/astro/src/core/create-vite.ts [hunk 1]
```diff
@@ -55,6 +55,7 @@ export async function createVite(inlineConfig: ViteConfigWithSSR, { astroConfig,
     ],
     publicDir: fileURLToPath(astroConfig.public),
     root: fileURLToPath(astroConfig.projectRoot),
+    envPrefix: 'PUBLIC_',
```
Comment (drwpow) [inline packages/astro/src/core/create-vite.ts:58]: 👍🏻

File: packages/astro/test/astro-envs.test.js [hunk 2]
```diff
@@ -0,0 +1,23 @@
+import { expect } from 'chai';
+import { loadFixture } from './test-utils.js';
+
+describe('Environment Variables', () => {
```
Comment (drwpow) [inline packages/astro/test/astro-envs.test.js:4]: Great test!

**Comment by jonathantneal** [comment]
> how _would_ you reference SECRET_PLACE if it were a DB key or something needed for the server?

This is an excellent question. Vite currently [discards them](https://github.com/vitejs/vite/blob/v2.6.13/packages/vite/src/node/config.ts#L1005-L1034). They could be used by other apps, but not Vite (and by extension Astro).

NextJS currently [preserves them on `process.env`](https://nextjs.org/docs/basic-features/environment-variables) (note: they have no `import.meta.env`), but they also [prevent them from being exposed to the client](https://nextjs.org/docs/basic-features/environment-variables#exposing-environment-variables-to-the-browser).

I’m not immediately sure how this should work in Astro, as nearly all of our code is technically used on the server.

**Review by drwpow — APPROVED** [review]
I think this is a great change. I want to +1 Nate’s question about env vars within Astro context vs client-side context. But I think that shouldn’t block this PR and is probably a larger discussion to have.

**Review by natemoo-re — APPROVED** [review]
Sounds good! Could you open an issue/RFC that we should resolve this discussion, @jonathantneal?

**Comment by jonathantneal** [comment]
@natemoo-re, yes, the RFC to access private variables is now available @ https://github.com/snowpackjs/astro/issues/1765

**Pull request merged**